### ISSUE TITLE
[stable/redis] redis container from previous master is not able to register as slave, if a password is used

### ISSUE THREAD

**Issue by avirtual** [helm/charts#14627]
I've installed one redis with 2 slaves. 
Master:  "10.240.13.208"
Slave0: "10.240.11.195:6379"
Slave1: "10.240.10.190:6379"

kubectl delete pod master-pod

Sentinels detect master pod is not responding and slave1 is promoted as master.
```
# +odown master mymaster 10.240.13.208 6379 #quorum 2/2
# +config-update-from sentinel f9b2d0c3c3a336761a130320872c48aacf9300d3 10.240.11.195 26379 @ mymaster 10.240.13.208 6379
# +switch-master mymaster 10.240.13.208 6379 10.240.10.190 6379

```

Master restarts and receives ip "10.240.13.209". It detects there is a master and assumes the role of a slave:

```
# +monitor master mymaster 10.240.13.209 6379 quorum 2
# +new-epoch 1
# +config-update-from sentinel da1881ed0cd2a59bd0abd3ec33abcb0057c23946 10.240.10.190 26379 @ mymaster 10.240.13.209 6379
# +switch-master mymaster 10.240.13.209 6379 10.240.10.190 6379
* +slave slave 10.240.13.209:6379 10.240.13.209 6379 @ mymaster 10.240.10.190 6379

```

Meanwhile the former slaves detect there is a new slave and supposedly register the new ip of the former master as a slave:

```
* +sentinel sentinel 13aa0ef3601b5ce25547ca0ce09eaa1f853ee641 10.240.13.209 26379 @ mymaster 10.240.10.190 6379
# +sdown slave 10.240.13.208:6379 10.240.13.208 6379 @ mymaster 10.240.10.190 6379

```

The problem is that when doing 
`kubectl exec -it pod-slave-1 -c sentinel -- redis-cli -h redis-svc -p 26379 sentinel slaves mymaster
`  you get (roughly 66% of the time) the old ip as a slave.

If you interogate the sentinels of the former slaves (one of them now a master) they will return 10.240.13.208 as a slave, even though they supposedly registered 10.240.13.209 as a new slave.
If you interogate 10.240.13.209, it will correctly return its own ip as a slave.

Any thoughts as to how to convince the sentinels to actually report the new ip of the registered slave?

**Comment by avirtual** [comment]
It would appear the problem is a missing line in templates/redis-master-statefulset.yaml
The slave-statefulsets have:
ARGS+=("--masterauth" "${REDIS_MASTER_PASSWORD}") which allows the redis container to connect to the redis master using the password.

The redis container from the redis-master-statefulset.yaml does not add it, probably under the assumption that the redis is a master and will not have to connect anywhere.
When the master redis respawns and gets a new ip, it register its sentinel container, but it fails to register the redis container as a slave, because it is missing the password.
Adding 
```
{{- if .Values.usePassword }}
          ARGS+=("--requirepass" "${REDIS_PASSWORD}")
>>>>>    ARGS+=("--masterauth" "${REDIS_PASSWORD}")
  {{- else }}
```
fixed my issue and now the redis container registers itself as a well behaving slave, after it unexpectedly dies.

**Comment by miguelaeh** [comment]
Hi @avirtual ,
How are you deploying the chart? I cannot reproduce that. I deploy the chart, and after that kill the master pod. What is happening in my case is that the slaves can't connect to the master, bu when the master is created again, the slaves synchronize with him. But any slave is promoting to master.

**Comment by avirtual** [comment]
Yes, i specified in the second comment. One slave is promoted to master, but upon resurrection the redis container from the initial 'master' pod is not registered due to the missing password.

As for deployment there's nothing special about it:
helms upgrade --install redis-${namespace:-test} ./ --namespace $namespace --debug --set redis.clusterDomain=kube-staging,redis.sentinel.clusterDomain=kube-staging

In short when the initial master pod comes back up, its sentinel is registering, but the redis container cannot do it, because it is missing "--masterauth".

**Comment by avirtual** [comment]
kubectl logs redis-nm-master-0 -n nm -c redis-nm
```
1:M 11 Jun 2019 07:03:52.608 * Ready to accept connections
1:S 11 Jun 2019 07:04:07.144 * Before turning into a replica, using my master parameters to synthesize a cached master: I may be able to synchronize with the new master with just a partial transfer.
1:S 11 Jun 2019 07:04:07.144 * REPLICAOF 10.240.3.21:6379 enabled (user request from 'id=5 addr=10.240.1.52:57445 fd=9 name=sentinel-f2674f4f-cmd age=10 idle=0 flags=x db=0 sub=0 psub=0 multi=3 qbuf=151 qbuf-free=32617 obl=36 oll=0 omem=0 events=r cmd=exec')
1:S 11 Jun 2019 07:04:07.144 # CONFIG REWRITE executed with success.
1:S 11 Jun 2019 07:04:07.635 * Connecting to MASTER 10.240.3.21:6379
1:S 11 Jun 2019 07:04:07.635 * MASTER <-> REPLICA sync started
1:S 11 Jun 2019 07:04:07.636 * Non blocking connect for SYNC fired the event.
1:S 11 Jun 2019 07:04:07.636 * Master replied to PING, replication can continue...
1:S 11 Jun 2019 07:04:07.637 * (Non critical) Master does not understand REPLCONF listening-port: -NOAUTH Authentication required.
1:S 11 Jun 2019 07:04:07.638 * (Non critical) Master does not understand REPLCONF capa: -NOAUTH Authentication required.
1:S 11 Jun 2019 07:04:07.638 * Trying a partial resynchronization (request 12e0b2609ee2779f26c7986055c51758a41fb31a:1).
1:S 11 Jun 2019 07:04:07.638 # Unexpected reply to PSYNC from master: -NOAUTH Authentication required.
1:S 11 Jun 2019 07:04:07.638 * Discarding previously cached master state.
1:S 11 Jun 2019 07:04:07.638 * Retrying with SYNC...
1:S 11 Jun 2019 07:04:07.639 # MASTER aborted replication with an error: NOAUTH Authentication required.
1:S 11 Jun 2019 07:04:08.637 * Connecting to MASTER 10.240.3.21:6379
1:S 11 Jun 2019 07:04:08.637 * MASTER <-> REPLICA sync started
1:S 11 Jun 2019 07:04:08.638 * Non blocking connect for SYNC fired the event.
1:S 11 Jun 2019 07:04:08.639 * Master replied to PING, replication can continue...
1:S 11 Jun 2019 07:04:08.639 * (Non critical) Master does not understand REPLCONF listening-port: -NOAUTH Authentication required.
1:S 11 Jun 2019 07:04:08.640 * (Non critical) Master does not understand REPLCONF capa: -NOAUTH Authentication required.
1:S 11 Jun 2019 07:04:08.640 * Partial resynchronization not possible (no cached master)

```

**Comment by miguelaeh** [comment]
Sorry I write wrong, I mean that, in my case, no slave is being promoted.
This is the master log:
```00000000000000000000000000000')
1:M 11 Jun 2019 06:46:21.777 * Waiting for end of BGSAVE for SYNC
1:M 11 Jun 2019 06:46:21.864 * Background saving terminated with success
1:M 11 Jun 2019 06:46:21.865 * Synchronization with replica 172.17.0.7:6379 succeeded
1:M 11 Jun 2019 06:46:21.865 * Synchronization with replica 172.17.0.6:6379 succeeded
1:M 11 Jun 2019 06:46:22.877 * Replica 172.17.0.10:6379 asks for synchronization
1:M 11 Jun 2019 06:46:22.877 * Partial resynchronization not accepted: Replication ID mismatch (Replica asked for 'c9b5909f57f8a91c8042ae1895ce5b2fa88f7ccc', my replication IDs are 'f460d1467b8ac86fa33a5feaa3e775ebc406ce0b' and '0000000000000000000000000000000000000000')
1:M 11 Jun 2019 06:46:22.877 * Starting BGSAVE for SYNC with target: disk
1:M 11 Jun 2019 06:46:22.878 * Background saving started by pid 25
25:C 11 Jun 2019 06:46:22.880 * DB saved on disk
25:C 11 Jun 2019 06:46:22.881 * RDB: 0 MB of memory used by copy-on-write
1:M 11 Jun 2019 06:46:22.974 * Background saving terminated with success
1:M 11 Jun 2019 06:46:22.974 * Synchronization with replica 172.17.0.10:6379 succeeded```

And the slave (all th slave has a similar log):
```1:S 11 Jun 2019 06:46:19.760 # Unable to connect to MASTER: Resource temporarily unavailable
1:S 11 Jun 2019 06:46:20.767 * Connecting to MASTER insipid-ibis-redis-master-0.insipid-ibis-redis-headless.default.svc.cluster.local:6379
1:S 11 Jun 2019 06:46:20.769 # Unable to connect to MASTER: Resource temporarily unavailable
1:S 11 Jun 2019 06:46:21.775 * Connecting to MASTER insipid-ibis-redis-master-0.insipid-ibis-redis-headless.default.svc.cluster.local:6379
1:S 11 Jun 2019 06:46:21.776 * MASTER <-> REPLICA sync started
1:S 11 Jun 2019 06:46:21.776 * Non blocking connect for SYNC fired the event.
1:S 11 Jun 2019 06:46:21.776 * Master replied to PING, replication can continue...
1:S 11 Jun 2019 06:46:21.777 * Trying a partial resynchronization (request c9b5909f57f8a91c8042ae1895ce5b2fa88f7ccc:435).
1:S 11 Jun 2019 06:46:21.777 * Full resync from master: f460d1467b8ac86fa33a5feaa3e775ebc406ce0b:0
1:S 11 Jun 2019 06:46:21.777 * Discarding previously cached master state.
1:S 11 Jun 2019 06:46:21.865 * MASTER <-> REPLICA sync: receiving 175 bytes from master
1:S 11 Jun 2019 06:46:21.866 * MASTER <-> REPLICA sync: Flushing old data
1:S 11 Jun 2019 06:46:21.866 * MASTER <-> REPLICA sync: Loading DB in memory
1:S 11 Jun 2019 06:46:21.866 * MASTER <-> REPLICA sync: Finished with success```

**Comment by avirtual** [comment]
are you enabling sentinel? slaves wont be promoted in the absence of sentinel.

**Comment by miguelaeh** [comment]
Hi @avirtual ,
Your are right, I forgot to enable sentinel. In any case executing this command:
`helm install stable/redis --set redis.clusterDomain=kube-staging,redis.sentinel.clusterDomain=kube-staging,sentinel.enabled=true`
And killing the master pod I am not able to  replicate your issue. You can see in this logs that the master is again the same:
```1:S 12 Jun 2019 07:25:48.058 * Background AOF rewrite finished successfully
1:S 12 Jun 2019 07:26:10.002 # Connection with master lost.
1:S 12 Jun 2019 07:26:10.002 * Caching the disconnected master state.
1:S 12 Jun 2019 07:26:10.002 * REPLICAOF 172.17.0.8:6379 enabled (user request from 'id=5 addr=172.17.0.8:38947 fd=8 name=sentinel-06752321-cmd age=20 idle=0 flags=x db=0 sub=0 psub=0 multi=3 qbuf=150 qbuf-free=32618 obl=36 oll=0 omem=0 events=r cmd=exec')
1:S 12 Jun 2019 07:26:10.003 # CONFIG REWRITE executed with success.
1:S 12 Jun 2019 07:26:10.038 * Connecting to MASTER 172.17.0.8:6379
1:S 12 Jun 2019 07:26:10.038 * MASTER <-> REPLICA sync started
1:S 12 Jun 2019 07:26:10.038 * Non blocking connect for SYNC fired the event.
1:S 12 Jun 2019 07:26:10.039 * Master replied to PING, replication can continue...
1:S 12 Jun 2019 07:26:10.040 * Trying a partial resynchronization (request e4f638f6edad74f4c763752c29c2f1d9e12797e6:2077).
1:S 12 Jun 2019 07:26:10.041 * Successful partial resynchronization with master.
1:S 12 Jun 2019 07:26:10.041 * MASTER <-> REPLICA sync: Master accepted a Partial Resynchronization.
1:S 12 Jun 2019 07:26:34.747 # Connection with master lost.
1:S 12 Jun 2019 07:26:34.747 * Caching the disconnected master state.
1:S 12 Jun 2019 07:26:35.218 * Connecting to MASTER 172.17.0.8:6379
1:S 12 Jun 2019 07:26:35.218 * MASTER <-> REPLICA sync started
1:S 12 Jun 2019 07:26:38.319 # Error condition on socket for SYNC: Connection refused
1:S 12 Jun 2019 07:26:39.252 * Connecting to MASTER 172.17.0.8:6379
1:S 12 Jun 2019 07:26:39.252 * MASTER <-> REPLICA sync started
1:S 12 Jun 2019 07:26:39.252 # Error condition on socket for SYNC: Connection refused
1:S 12 Jun 2019 07:26:40.259 * Connecting to MASTER 172.17.0.8:6379
1:S 12 Jun 2019 07:26:40.259 * MASTER <-> REPLICA sync started
1:S 12 Jun 2019 07:26:40.259 * Non blocking connect for SYNC fired the event.
1:S 12 Jun 2019 07:26:40.259 * Master replied to PING, replication can continue...
```

Also, the sentinel logs detect the master has down, but there are no master change.  Did you deploy in the same way as me? Or do you perform more configuration changes on the deployment from the default?

**Comment by avirtual** [comment]
My guess is that your master is not down long enough for the sentinels to decide the master is down. 
I've changed the default sentinel values for:
```
downAfterMilliseconds: 6000
failoverTimeout: 10000
```

The whole purpose of the sentinels is to detect a master is dead and replace it, so if your sentinels are not doing that, i wouldn't say it's something desirable :)

**Comment by miguelaeh** [comment]
Hi @avirtual ,
Thanks, you are right. 
I replicated it. And also it seems your solution works. But in my case, none of the pods died unexpectedly. And the master starts as slave successfully and connect with the new master:
```
1:C 13 Jun 2019 09:23:38.516 # oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
1:C 13 Jun 2019 09:23:38.516 # Redis version=5.0.5, bits=64, commit=00000000, modified=0, pid=1, just started
1:C 13 Jun 2019 09:23:38.516 # Configuration loaded
1:M 13 Jun 2019 09:23:38.518 * Running mode=standalone, port=6379.
1:M 13 Jun 2019 09:23:38.518 # WARNING: The TCP backlog setting of 511 cannot be enforced because /proc/sys/net/core/somaxconn is set to the lower value of 128.
1:M 13 Jun 2019 09:23:38.518 # Server initialized
1:M 13 Jun 2019 09:23:38.518 * Ready to accept connections
1:S 13 Jun 2019 09:23:49.038 * Before turning into a replica, using my master parameters to synthesize a cached master: I may be able to synchronize with the new master with just a partial transfer.
1:S 13 Jun 2019 09:23:49.038 * REPLICAOF 172.17.0.9:6379 enabled (user request from 'id=5 addr=172.17.0.9:35993 fd=9 name=sentinel-94b1ccb5-cmd age=11 idle=0 flags=x db=0 sub=0 psub=0 multi=3 qbuf=150 qbuf-free=32618 obl=36 oll=0 omem=0 events=r cmd=exec')
1:S 13 Jun 2019 09:23:49.038 # CONFIG REWRITE executed with success.
1:S 13 Jun 2019 09:23:49.904 * Connecting to MASTER 172.17.0.9:6379
1:S 13 Jun 2019 09:23:49.905 * MASTER <-> REPLICA sync started
1:S 13 Jun 2019 09:23:49.905 * Non blocking connect for SYNC fired the event.
1:S 13 Jun 2019 09:23:49.906 * Master replied to PING, replication can continue...
1:S 13 Jun 2019 09:23:49.906 * Trying a partial resynchronization (request e70f8b37096e84a8e17b864390989e5ab3083471:1).
1:S 13 Jun 2019 09:23:49.907 * Full resync from master: 58a7380346dc11b794c1fa6adc36663bcba3ac75:30938
1:S 13 Jun 2019 09:23:49.908 * Discarding previously cached master state.
1:S 13 Jun 2019 09:23:50.005 * MASTER <-> REPLICA sync: receiving 176 bytes from master
1:S 13 Jun 2019 09:23:50.005 * MASTER <-> REPLICA sync: Flushing old data
1:S 13 Jun 2019 09:23:50.006 * MASTER <-> REPLICA sync: Loading DB in memory
1:S 13 Jun 2019 09:23:50.006 * MASTER <-> REPLICA sync: Finished with success
1:S 13 Jun 2019 09:23:50.006 * Background append only file rewriting started by pid 42
1:S 13 Jun 2019 09:23:50.044 * AOF rewrite child asks to stop sending diffs.
42:C 13 Jun 2019 09:23:50.044 * Parent agreed to stop sending diffs. Finalizing AOF...
42:C 13 Jun 2019 09:23:50.045 * Concatenating 0.00 MB of AOF diff received from parent.
42:C 13 Jun 2019 09:23:50.045 * SYNC append only file rewrite performed
42:C 13 Jun 2019 09:23:50.046 * AOF rewrite: 0 MB of memory used by copy-on-write
1:S 13 Jun 2019 09:23:50.105 * Background AOF rewrite terminated with success
1:S 13 Jun 2019 09:23:50.105 * Residual parent diff successfully flushed to the rewritten AOF (0.00 MB)
1:S 13 Jun 2019 09:23:50.105 * Background AOF rewrite finished successfully
```

Those are the master logs after kill him and restart.
The following are the logs of the slave that promote to master:
```1:C 13 Jun 2019 09:21:22.188 # oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
1:C 13 Jun 2019 09:21:22.188 # Redis version=5.0.5, bits=64, commit=00000000, modified=0, pid=1, just started
1:C 13 Jun 2019 09:21:22.188 # Configuration loaded
1:S 13 Jun 2019 09:21:22.189 * Running mode=standalone, port=6379.
1:S 13 Jun 2019 09:21:22.190 # WARNING: The TCP backlog setting of 511 cannot be enforced because /proc/sys/net/core/somaxconn is set to the lower value of 128.
1:S 13 Jun 2019 09:21:22.190 # Server initialized
1:S 13 Jun 2019 09:21:22.190 * Ready to accept connections
1:S 13 Jun 2019 09:21:22.190 * Connecting to MASTER dozing-moth-redis-master-0.dozing-moth-redis-headless.default.svc.cluster.local:6379
1:S 13 Jun 2019 09:21:22.192 * MASTER <-> REPLICA sync started
1:S 13 Jun 2019 09:21:22.192 * Non blocking connect for SYNC fired the event.
1:S 13 Jun 2019 09:21:22.192 * Master replied to PING, replication can continue...
1:S 13 Jun 2019 09:21:22.192 * Partial resynchronization not possible (no cached master)
1:S 13 Jun 2019 09:21:22.193 * Full resync from master: afe320ed107b6f94bf0ae67b9de2a82ce136de46:2481
1:S 13 Jun 2019 09:21:22.276 * MASTER <-> REPLICA sync: receiving 176 bytes from master
1:S 13 Jun 2019 09:21:22.276 * MASTER <-> REPLICA sync: Flushing old data
1:S 13 Jun 2019 09:21:22.276 * MASTER <-> REPLICA sync: Loading DB in memory
1:S 13 Jun 2019 09:21:22.276 * MASTER <-> REPLICA sync: Finished with success
1:S 13 Jun 2019 09:21:22.277 * Background append only file rewriting started by pid 15
1:S 13 Jun 2019 09:21:22.307 * AOF rewrite child asks to stop sending diffs.
15:C 13 Jun 2019 09:21:22.308 * Parent agreed to stop sending diffs. Finalizing AOF...
15:C 13 Jun 2019 09:21:22.308 * Concatenating 0.00 MB of AOF diff received from parent.
15:C 13 Jun 2019 09:21:22.308 * SYNC append only file rewrite performed
15:C 13 Jun 2019 09:21:22.309 * AOF rewrite: 0 MB of memory used by copy-on-write
1:S 13 Jun 2019 09:21:22.394 * Background AOF rewrite terminated with success
1:S 13 Jun 2019 09:21:22.394 * Residual parent diff successfully flushed to the rewritten AOF (0.00 MB)
1:S 13 Jun 2019 09:21:22.394 * Background AOF rewrite finished successfully
1:S 13 Jun 2019 09:21:44.374 # Connection with master lost.
1:S 13 Jun 2019 09:21:44.374 * Caching the disconnected master state.
1:S 13 Jun 2019 09:21:44.374 * REPLICAOF 172.17.0.3:6379 enabled (user request from 'id=4 addr=172.17.0.9:44465 fd=8 name=sentinel-94b1ccb5-cmd age=20 idle=0 flags=x db=0 sub=0 psub=0 multi=3 qbuf=150 qbuf-free=32618 obl=36 oll=0 omem=0 events=r cmd=exec')
1:S 13 Jun 2019 09:21:44.375 # CONFIG REWRITE executed with success.
1:S 13 Jun 2019 09:21:45.342 * Connecting to MASTER 172.17.0.3:6379
1:S 13 Jun 2019 09:21:45.342 * MASTER <-> REPLICA sync started
1:S 13 Jun 2019 09:21:45.342 * Non blocking connect for SYNC fired the event.
1:S 13 Jun 2019 09:21:45.342 * Master replied to PING, replication can continue...
1:S 13 Jun 2019 09:21:45.343 * Trying a partial resynchronization (request afe320ed107b6f94bf0ae67b9de2a82ce136de46:6732).
1:S 13 Jun 2019 09:21:45.344 * Successful partial resynchronization with master.
1:S 13 Jun 2019 09:21:45.344 * MASTER <-> REPLICA sync: Master accepted a Partial Resynchronization.
1:S 13 Jun 2019 09:23:33.511 # Connection with master lost.
1:S 13 Jun 2019 09:23:33.511 * Caching the disconnected master state.
1:S 13 Jun 2019 09:23:34.099 * Connecting to MASTER 172.17.0.3:6379
1:S 13 Jun 2019 09:23:34.099 * MASTER <-> REPLICA sync started
1:M 13 Jun 2019 09:23:36.953 # Setting secondary replication ID to afe320ed107b6f94bf0ae67b9de2a82ce136de46, valid up to offset: 28607. New replication ID is 58a7380346dc11b794c1fa6adc36663bcba3ac75
1:M 13 Jun 2019 09:23:36.953 * Discarding previously cached master state.
1:M 13 Jun 2019 09:23:36.953 * MASTER MODE enabled (user request from 'id=12 addr=172.17.0.8:56239 fd=9 name=sentinel-51c104d0-cmd age=112 idle=0 flags=x db=0 sub=0 psub=0 multi=3 qbuf=140 qbuf-free=32628 obl=36 oll=0 omem=0 events=r cmd=exec')
1:M 13 Jun 2019 09:23:36.955 # CONFIG REWRITE executed with success.
1:M 13 Jun 2019 09:23:37.972 * Replica 172.17.0.8:6379 asks for synchronization
1:M 13 Jun 2019 09:23:37.972 * Partial resynchronization request from 172.17.0.8:6379 accepted. Sending 293 bytes of backlog starting from offset 28607.
1:M 13 Jun 2019 09:23:49.907 * Replica 172.17.0.3:6379 asks for synchronization
1:M 13 Jun 2019 09:23:49.907 * Partial resynchronization not accepted: Replication ID mismatch (Replica asked for 'e70f8b37096e84a8e17b864390989e5ab3083471', my replication IDs are '58a7380346dc11b794c1fa6adc36663bcba3ac75' and 'afe320ed107b6f94bf0ae67b9de2a82ce136de46')
1:M 13 Jun 2019 09:23:49.907 * Starting BGSAVE for SYNC with target: disk
1:M 13 Jun 2019 09:23:49.907 * Background saving started by pid 251
251:C 13 Jun 2019 09:23:49.909 * DB saved on disk
251:C 13 Jun 2019 09:23:49.910 * RDB: 0 MB of memory used by copy-on-write
1:M 13 Jun 2019 09:23:50.005 * Background saving terminated with success
1:M 13 Jun 2019 09:23:50.005 * Synchronization with replica 172.17.0.3:6379 succeeded
```

As you can see, the old master was 172.17.0.3 and now it connects to the new master as slave. I guess that is what you expected, right?

**Comment by avirtual** [comment]
Yes, it is expected for the master to reconnect as a slave, but it won't do it without the bit that adds --masterauth.
Having dealt with redis and sentinel for a couple of days, I feel there are some issues that would have to be addressed.
Quorum is used to decide a master is dead, but it is not used to elect a new master. This essentially means that if you start with a master and 2 slaves, the system will survive to one unexpected crash, but not if the new master crashes again.
Electing a new master requires sentinels_no/2 + 1.
After master restarts, its old sentinel will be counted, even though it is inaccessible, so you will have 3 alive sentinels and one dead. If the new master dies, you're out of luck, there is quorum to decide a master is dead, but not a new master won't be elected because of the dead sentinel in the pool.

**Comment by miguelaeh** [comment]
Hi @avirtual ,
First, you are right we have to add the `masterauth` parameter to the master.
In the other hand, how did you test that:
>After master restarts, its old sentinel will be counted, even though it is inaccessible, so you will have 3 alive sentinels and one dead. If the new master dies, you're out of luck, there is quorum to decide a master is dead, but a new master won't be elected because of the dead sentinel in the pool.

I can delete the master multiple times and a new master is elected always. The problem comes when you delete two of the 3 pods. That is because more than a half of the cluster has died, so the cluster is in a inconsistent state.
Also, in case sentinels are on a pool, as the chart is using statefulsets the new sentinel will have the same name as the one that died, so I guess it is like if it where the same again.

**Comment by avirtual** [comment]
After you delete the master, it register itself as a slave, so if you delete that pod again, you'll actually be deleting a slave, not the new master. 
By master i mean whichever pod is considered to be the master of the other 2 redis containers. (initially it is the pod named redis-namespace-master-0, but then it could be redis-namespace-slave-0 or redis-namespace-slave-1)

The scenario is:
1. delete the master (pod -master-0)
1.a one of the slaves becomes master
1.b the former master respawns and registers as a slave

2. (after the former master is up and running as a slave) identify which of the former slaves became master and delete its pod. Normally the process of sentinels electing a new master should be the same as the first time. The 2 alive sentinels meet the quorum and decide that the new master has indeed died. The problem is that a new master will not be elected this time, because there are 2 dead sentinels.

So in fact you do not delete 2 pods at the same time, you just do it one by one, expecting the system to survive indefinitely, as at any point you'd have one promoted master and 2 existing slaves.

**Comment by avirtual** [comment]
To summarize, the pull request will fix the bug report but, in my opinion, the process of electing a new master needs to be fixed (probably by not counting the dead sentinels)

**Comment by miguelaeh** [comment]
Hi, 
>identify which of the former slaves became master and delete its pod.

that is what I did. But in my case I found no problem.
